src/js/program/program-controller.js

```diff
@@ -4,34 +4,61 @@ import cancelable from 'utils/cancelable';
 import { MediaControllerListener } from 'program/program-listeners';
 import Events from 'utils/backbone.events';
 import BackgroundMedia from 'program/background-media';
-import { PLAYER_STATE, STATE_BUFFERING, STATE_IDLE, STATE_PAUSED } from 'events/events';
+import { PLAYER_STATE, STATE_IDLE, STATE_BUFFERING, STATE_PAUSED } from 'events/events';
 import { PlayerError, MSG_CANT_PLAY_VIDEO, ERROR_PLAYLIST_ITEM_MISSING_SOURCE } from 'api/errors';
+import { ItemPromise } from 'controller/next-item-promise';
+import { normalizePlaylistItem } from 'playlist/playlist';
+import Item from 'playlist/item';
 
 class ProgramController extends Events {
     /**
      * ProgramController constructor
      * @param {Model} model - The player's model
      * @param {MediaElementPool} mediaPool - The player's media element pool
+     * @param {Api} api - The player API applied as context in async playlist item promise callback execution
      */
-    constructor(model, mediaPool) {
+    constructor(model, mediaPool, api) {
         super();
 
         this.adPlaying = false;
+        this.api = api;
         this.background = BackgroundMedia();
         this.mediaPool = mediaPool;
         this.mediaController = null;
         this.mediaControllerListener = MediaControllerListener(model, this);
         this.model = model;
         this.providers = new Providers(model.getConfiguration());
-        this.loadPromise = Promise.resolve();
+        this.loadPromise = null;
         this.backgroundLoading = model.get('backgroundLoading');
+        this.itemPromises = [];
 
         if (!this.backgroundLoading) {
             // If background loading is not supported, set the shared media element
             model.set('mediaElement', this.mediaPool.getPrimedElement());
         }
     }
 
+    asyncItem(index, item) {
+        const { api, model } = this;
+        // Set the player state to buffering if there is a playlist item callback
+        const deferBufferingState = setTimeout(() => {
+            model.set(PLAYER_STATE, STATE_BUFFERING);
+        });
+        return this.getItemPromise(index, api).run().then((playlistItem) => {
+            clearTimeout(deferBufferingState);
+            if (playlistItem && playlistItem !== item) {
+                this.setPlaylistItem(index, playlistItem);
```

Comment:
`this.setPlaylistItem(index, playlistItem)` creates a new instance of `Item` to set it to `playlist[index]`, but `playlistItem` is being used to set to `this.item` in `media-controller`.

DAI plugin uses `getPlaylistItem` to get the current playlistItem and modify the source, but hls uses the item in `media-controller`.

Can we change this to `return this.setPlaylistItem(index, playlistItem);` so that the active `item` in `media-controller` matches the result of `getPlaylistItem`?
If not, is there another API that DAI plugin can use to get the active item in `media-controller`?

Reply:
@pajong 
Once the media-controller is created for an item and active (on `"playlistItem"` event), `mediaController.item` is the active item loaded into the provider. You can change the item in the playlist, but it will not change the active item. That hasn't changed.

The `"playlistItem"`  event should contain the active item. The `getPlaylistItem(index)` method returns what is in the playlist.

If you change what is in the playlist after "playlistItem" with `setPlaylistItem`, that change is not active until navigating back to that item index.

Please describe what if any test case is affected. Include a link, repro steps, and expected/actual behavior. It would help if you also include links to the implementation in DAI for clarity.

If there isn't a new bug, and you are just concerned with how the async API could impact DAI functionality, please provide examples with links to DAI code.

Reply:
The problem that I am seeing with DAI is that `this.setPlaylistItem(index, playlistItem)` creates a new instance, then sets to `playlist`, but `playlistItem` is used to set in the `media-controller` after this point. My question is simply why we are not setting it to the newly created instance if we are setting `mediaController.item` anyways.

There is no current test pages broken, but the problem only surfaces if we have the `playlistItemPromise` -  I can put up a test page to reflect that if that helps

Reply:
Start with linking to the code in DAI. I think your point is more about when this code reads the item, and that `api.setPlaylistItem` can be called externally while one or more of these promises is unresolved.